rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs

```diff
@@ -97,90 +97,264 @@ impl From<(u64, Duration)> for Throughput {
+    /// to track throughput. The time window should configured to be large enough
+    /// so that these sub-windows aren't too small for network-based events.
+    /// A time window of 10ms probably won't work, but 500ms might. The default
+    /// is one second.
+    pub(super) fn new(time_window: Duration, now: SystemTime) -> Self {
+        assert!(!time_window.is_zero());
+        let resolution = time_window.div_f64(BUFFER_SIZE as f64);
+        Self {
+            resolution,
+            current_tail: now,
+            buffer: LogBuffer::new(),
+        }
+    }
+
+    /// Returns the resolution at which events are logged at.
+    pub(super) fn resolution(&self) -> Duration {
+        self.resolution
+    }
+
+    /// Pushes a "pending" event.
+    ///
+    /// Pending indicates the streaming future is waiting for something.
+    /// In an upload, it is waiting for data from the user, and in a download,
+    /// it is waiting for data from the server.
+    pub(super) fn push_pending(&mut self, time: SystemTime) {
+        self.push(time, Cell::Pending(0));
+    }
+
+    /// Pushes a data transferred event.
+    ///
+    /// Indicates that this number of bytes were transferred at this time.
+    pub(super) fn push_bytes_transferred(&mut self, time: SystemTime, bytes: u64) {
+        self.push(time, Cell::TransferredBytes(bytes));
+    }
+
+    fn push(&mut self, now: SystemTime, value: Cell) {
+        self.catch_up(now);
+        self.buffer.tail_mut().merge(value);
+        self.buffer.fill_gaps();
+    }
+
+    /// Pushes empty cells until `current_tail` is caught up to `now`.
+    fn catch_up(&mut self, now: SystemTime) {
+        while now >= self.current_tail {
+            self.current_tail += self.resolution;
+            self.buffer.push(Cell::Empty);
+        }
+        assert!(self.current_tail >= now);
+    }
+
+    /// Generates an overall report of the time window.
+    pub(super) fn report(&mut self, now: SystemTime) -> ThroughputReport {
+        self.catch_up(now);
+        self.buffer.fill_gaps();
+
+        let CellCounts {
+            empty,
+            no_polling,
+            transferred,
+            pending,
+        } = self.buffer.counts();
+
+        let bytes = self.buffer.bytes_transferred();
+        let time = self.resolution * (BUFFER_SIZE - empty) as u32;
+        let throughput = Throughput::new(bytes, time);
+
+        let half = BUFFER_SIZE / 2;
+        if empty >= half {
+            return ThroughputReport::Incomplete;
+        }
+        match (transferred > 0, no_polling >= half, pending >= half) {
+            (true, _, _) => ThroughputReport::Transferred(throughput),
+            (_, true, _) => ThroughputReport::NoPolling,
+            (_, _, true) => ThroughputReport::Pending,
+            _ => ThroughputReport::Incomplete,
+        }
     }
 }
 
 #[cfg(test)]
 mod test {
-    use super::{Throughput, ThroughputLogs};
-    use std::time::{Duration, SystemTime, UNIX_EPOCH};
+    use super::*;
```

Comment:
lol I'm always making these explicit so that we don't import things in the parent scope that only get used in the child scope. We're covering each other's paint.

Reply:
I don't want to think about imports when writing tests 😆